### ISSUE TITLE
Some errors on installing an extension 

### ISSUE THREAD

**Issue by salarmehr-cbs** [FriendsOfPHP/pickle#175]
(related to https://github.com/FriendsOfPHP/pickle/issues/162) 

After `git pull` and `composer update` I see that now it progress more with the installation.
It actually updates the php.ini file and creates the file in the ext folder. however, it also shows the following errors

```
C:\server\pickle>php bin/pickle install memcache
+---------------+--------------------------+
| php Path      | C:^\server^\php^\php.exe |
| php Version   | 7.4.5                    |
| Compiler      | vc15                     |
| Architecture  | x64                      |
| Thread safety | no                       |
| Extension dir | C:\server\php\ext        |
| php.ini       | C:\server\php\php.ini    |
+---------------+--------------------------+
downloading https://windows.php.net/downloads/pecl/releases/memcache/4.0.5.2/php_memcache-4.0.5.2-7.4-nts-vc15-x64.zip
 223981/231863 [===========================>]  96%PHP Warning:  file_get_contents(): SSL operation failed with code 1. OpenSSL Error messages:
error:14095126:SSL routines:ssl3_read_n:unexpected eof while reading in C:\server\pickle\src\Package\PHP\Command\Install\Windows\Binary.php on line 211
PHP Stack trace:
PHP   1. {main}() C:\server\pickle\bin\pickle:0
PHP   2. Pickle\Console\Application->run() C:\server\pickle\bin\pickle:39
PHP   3. Pickle\Console\Application->doRun() C:\server\pickle\vendor\symfony\console\Application.php:117
PHP   4. Pickle\Console\Application->doRunCommand() C:\server\pickle\vendor\symfony\console\Application.php:185
PHP   5. Pickle\Console\Command\InstallerCommand->run() C:\server\pickle\vendor\symfony\console\Application.php:835
PHP   6. Pickle\Console\Command\InstallerCommand->execute() C:\server\pickle\vendor\symfony\console\Command\Command.php:245
PHP   7. Pickle\Console\Command\InstallerCommand->binaryInstallWindows() C:\server\pickle\src\Console\Command\InstallerCommand.php:203
PHP   8. Pickle\Package\PHP\Command\Install\Windows\Binary->install() C:\server\pickle\src\Console\Command\InstallerCommand.php:137
PHP   9. Pickle\Package\PHP\Command\Install\Windows\Binary->download() C:\server\pickle\src\Package\PHP\Command\Install\Windows\Binary.php:310
PHP  10. file_get_contents() C:\server\pickle\src\Package\PHP\Command\Install\Windows\Binary.php:211

Warning: file_get_contents(): SSL operation failed with code 1. OpenSSL Error messages:
error:14095126:SSL routines:ssl3_read_n:unexpected eof while reading in C:\server\pickle\src\Package\PHP\Command\Install\Windows\Binary.php on line 211

Call Stack:
    0.0008     396008   1. {main}() C:\server\pickle\bin\pickle:0
    0.0167     974832   2. Pickle\Console\Application->run() C:\server\pickle\bin\pickle:39
    0.0266    1291856   3. Pickle\Console\Application->doRun() C:\server\pickle\vendor\symfony\console\Application.php:117
    0.0476    2034048   4. Pickle\Console\Application->doRunCommand() C:\server\pickle\vendor\symfony\console\Application.php:185
    0.0476    2034048   5. Pickle\Console\Command\InstallerCommand->run() C:\server\pickle\vendor\symfony\console\Application.php:835
    0.0480    2039560   6. Pickle\Console\Command\InstallerCommand->execute() C:\server\pickle\vendor\symfony\console\Command\Command.php:245
    0.2289    2084224   7. Pickle\Console\Command\InstallerCommand->binaryInstallWindows() C:\server\pickle\src\Console\Command\InstallerCommand.php:203
    0.4782    2221280   8. Pickle\Package\PHP\Command\Install\Windows\Binary->install() C:\server\pickle\src\Console\Command\InstallerCommand.php:137
    9.4913    2222184   9. Pickle\Package\PHP\Command\Install\Windows\Binary->download() C:\server\pickle\src\Package\PHP\Command\Install\Windows\Binary.php:310
    9.4915    2223808  10. file_get_contents() C:\server\pickle\src\Package\PHP\Command\Install\Windows\Binary.php:211


Extracting archives...
copying C:\Users\Reza Salarmehr\AppData\Local\Temp\1\pickle-memcache\php_memcache.dll to C:\server\php\ext\php_memcache.dll



  [Exception]
  Cannot copy DLL <C:\Users\Reza Salarmehr\AppData\Local\Temp\1\pickle-memcache\php_memcache.dll> to <C:\server\php\ext\php_memcache.dll>
```

**Comment by pierrejoye** [comment]
Thanks for the report :)

It may be related to CA Cert or some SSL error.

Does it happen always when you run pecl install memcache?

**Comment by pierrejoye** [comment]
I commited a fix about the PHP core kind BC breaks. We lost support on windows for anyone doing custom php builds and not with VC15, but for any normal users, it works. I force VC to VC15 for now (all supported php versions are built using VC15 only).

**Comment by salarmehr-cbs** [comment]
@pierrejoye Thanks for what you did. Still, I got error:

```
C:\server\pickle>php bin/pickle install memcache
+---------------+-----------------------+
| php Path      | C:\server\php\php.exe |
| php Version   | 7.4.5                 |
| Compiler      | vc15                  |
| Architecture  | x64                   |
| Thread safety | no                    |
| Extension dir | C:\server\php\ext     |
| php.ini       | C:\server\php\php.ini |
+---------------+-----------------------+
downloading https://windows.php.net/downloads/pecl/releases/memcache/4.0.5.2/php_memcache-4.0.5.2-7.4-nts-vc15-x64.zip
 213302/231863 [=========================>--]  91%PHP Warning:  file_get_contents(): SSL operation failed with code 1. OpenSSL Error messages:
error:14095126:SSL routines:ssl3_read_n:unexpected eof while reading in C:\server\pickle\src\Package\PHP\Command\Install\Windows\Binary.php on line 229
PHP Stack trace:
PHP   1. {main}() C:\server\pickle\bin\pickle:0
PHP   2. Pickle\Console\Application->run() C:\server\pickle\bin\pickle:39
PHP   3. Pickle\Console\Application->doRun() C:\server\pickle\vendor\symfony\console\Application.php:140
PHP   4. Pickle\Console\Application->doRunCommand() C:\server\pickle\vendor\symfony\console\Application.php:264
PHP   5. Pickle\Console\Command\InstallerCommand->run() C:\server\pickle\vendor\symfony\console\Application.php:911
PHP   6. Pickle\Console\Command\InstallerCommand->execute() C:\server\pickle\vendor\symfony\console\Command\Command.php:258
PHP   7. Pickle\Console\Command\InstallerCommand->binaryInstallWindows() C:\server\pickle\src\Console\Command\InstallerCommand.php:200
PHP   8. Pickle\Package\PHP\Command\Install\Windows\Binary->install() C:\server\pickle\src\Console\Command\InstallerCommand.php:138
PHP   9. Pickle\Package\PHP\Command\Install\Windows\Binary->download() C:\server\pickle\src\Package\PHP\Command\Install\Windows\Binary.php:328
PHP  10. file_get_contents() C:\server\pickle\src\Package\PHP\Command\Install\Windows\Binary.php:229

Warning: file_get_contents(): SSL operation failed with code 1. OpenSSL Error messages:
error:14095126:SSL routines:ssl3_read_n:unexpected eof while reading in C:\server\pickle\src\Package\PHP\Command\Install\Windows\Binary.php on line 229

Call Stack:
    0.0010     396008   1. {main}() C:\server\pickle\bin\pickle:0
    0.0217    1143096   2. Pickle\Console\Application->run() C:\server\pickle\bin\pickle:39
    0.1038    1425832   3. Pickle\Console\Application->doRun() C:\server\pickle\vendor\symfony\console\Application.php:140
    0.1166    1959360   4. Pickle\Console\Application->doRunCommand() C:\server\pickle\vendor\symfony\console\Application.php:264
    0.1167    1959360   5. Pickle\Console\Command\InstallerCommand->run() C:\server\pickle\vendor\symfony\console\Application.php:911
    0.1169    1961456   6. Pickle\Console\Command\InstallerCommand->execute() C:\server\pickle\vendor\symfony\console\Command\Command.php:258
    0.1176    1965440   7. Pickle\Console\Command\InstallerCommand->binaryInstallWindows() C:\server\pickle\src\Console\Command\InstallerCommand.php:200
    0.6746    2415464   8. Pickle\Package\PHP\Command\Install\Windows\Binary->install() C:\server\pickle\src\Console\Command\InstallerCommand.php:138
    9.7893    2416368   9. Pickle\Package\PHP\Command\Install\Windows\Binary->download() C:\server\pickle\src\Package\PHP\Command\Install\Windows\Binary.php:328
    9.7903    2417992  10. file_get_contents() C:\server\pickle\src\Package\PHP\Command\Install\Windows\Binary.php:229

 231863/231863 [============================] 100%Extracting archives...
copying C:\Users\Reza Salarmehr\AppData\Local\Temp\1\pickle-memcache\php_memcache.dll to C:\server\php\ext\php_memcache.dll

PHP Warning:  copy(C:\server\php\ext/php_memcache.dll): failed to open stream: Resource temporarily unavailable in C:\server\pickle\src\Package\PHP\Command\Install\Windows\Binary.php on line 257
PHP Stack trace:
PHP   1. {main}() C:\server\pickle\bin\pickle:0
PHP   2. Pickle\Console\Application->run() C:\server\pickle\bin\pickle:39
PHP   3. Pickle\Console\Application->doRun() C:\server\pickle\vendor\symfony\console\Application.php:140
PHP   4. Pickle\Console\Application->doRunCommand() C:\server\pickle\vendor\symfony\console\Application.php:264
PHP   5. Pickle\Console\Command\InstallerCommand->run() C:\server\pickle\vendor\symfony\console\Application.php:911
PHP   6. Pickle\Console\Command\InstallerCommand->execute() C:\server\pickle\vendor\symfony\console\Command\Command.php:258
PHP   7. Pickle\Console\Command\InstallerCommand->binaryInstallWindows() C:\server\pickle\src\Console\Command\InstallerCommand.php:200
PHP   8. Pickle\Package\PHP\Command\Install\Windows\Binary->install() C:\server\pickle\src\Console\Command\InstallerCommand.php:138
PHP   9. Pickle\Package\PHP\Command\Install\Windows\Binary->copyFiles() C:\server\pickle\src\Package\PHP\Command\Install\Windows\Binary.php:330
PHP  10. copy() C:\server\pickle\src\Package\PHP\Command\Install\Windows\Binary.php:257

Warning: copy(C:\server\php\ext/php_memcache.dll): failed to open stream: Resource temporarily unavailable in C:\server\pickle\src\Package\PHP\Command\Install\Windows\Binary.php on line 257

Call Stack:
    0.0010     396008   1. {main}() C:\server\pickle\bin\pickle:0
    0.0217    1143096   2. Pickle\Console\Application->run() C:\server\pickle\bin\pickle:39
    0.1038    1425832   3. Pickle\Console\Application->doRun() C:\server\pickle\vendor\symfony\console\Application.php:140
    0.1166    1959360   4. Pickle\Console\Application->doRunCommand() C:\server\pickle\vendor\symfony\console\Application.php:264
    0.1167    1959360   5. Pickle\Console\Command\InstallerCommand->run() C:\server\pickle\vendor\symfony\console\Application.php:911
    0.1169    1961456   6. Pickle\Console\Command\InstallerCommand->execute() C:\server\pickle\vendor\symfony\console\Command\Command.php:258
    0.1176    1965440   7. Pickle\Console\Command\InstallerCommand->binaryInstallWindows() C:\server\pickle\src\Console\Command\InstallerCommand.php:200
    0.6746    2415464   8. Pickle\Package\PHP\Command\Install\Windows\Binary->install() C:\server\pickle\src\Console\Command\InstallerCommand.php:138
   13.7948    2419720   9. Pickle\Package\PHP\Command\Install\Windows\Binary->copyFiles() C:\server\pickle\src\Package\PHP\Command\Install\Windows\Binary.php:330
   13.8072    2420856  10. copy() C:\server\pickle\src\Package\PHP\Command\Install\Windows\Binary.php:257


In Binary.php line 259:

  Cannot copy DLL <C:\Users\Reza Salarmehr\AppData\Local\Temp\1\pickle-memcache\php_memcache.dll> to <C:\server\php\ext\php_memcache.dll>


install [--no-convert] [--defaults] [--source] [--with-configure-options WITH-CONFIGURE-OPTIONS] [--save-logs SAVE-LOGS] [--dry-run] [--php PHP] [--ini INI] [--tmp-dir TMP-DIR] [--binary] [--] [<path>]
```

**Comment by salarmehr-cbs** [comment]
If I delete the DLL and also remove its reference from the php.ini I will just get

```
C:\server\pickle>php bin/pickle install memcache
+---------------+-----------------------+
| php Path      | C:\server\php\php.exe |
| php Version   | 7.4.5                 |
| Compiler      | vc15                  |
| Architecture  | x64                   |
| Thread safety | no                    |
| Extension dir | C:\server\php\ext     |
| php.ini       | C:\server\php\php.ini |
+---------------+-----------------------+
downloading https://windows.php.net/downloads/pecl/releases/memcache/4.0.5.2/php_memcache-4.0.5.2-7.4-nts-vc15-x64.zip
 213302/231863 [=========================>--]  91%PHP Warning:  file_get_contents(): SSL operation failed with code 1. OpenSSL Error messages:
error:14095126:SSL routines:ssl3_read_n:unexpected eof while reading in C:\server\pickle\src\Package\PHP\Command\Install\Windows\Binary.php on line 229
PHP Stack trace:
PHP   1. {main}() C:\server\pickle\bin\pickle:0
PHP   2. Pickle\Console\Application->run() C:\server\pickle\bin\pickle:39
PHP   3. Pickle\Console\Application->doRun() C:\server\pickle\vendor\symfony\console\Application.php:140
PHP   4. Pickle\Console\Application->doRunCommand() C:\server\pickle\vendor\symfony\console\Application.php:264
PHP   5. Pickle\Console\Command\InstallerCommand->run() C:\server\pickle\vendor\symfony\console\Application.php:911
PHP   6. Pickle\Console\Command\InstallerCommand->execute() C:\server\pickle\vendor\symfony\console\Command\Command.php:258
PHP   7. Pickle\Console\Command\InstallerCommand->binaryInstallWindows() C:\server\pickle\src\Console\Command\InstallerCommand.php:200
PHP   8. Pickle\Package\PHP\Command\Install\Windows\Binary->install() C:\server\pickle\src\Console\Command\InstallerCommand.php:138
PHP   9. Pickle\Package\PHP\Command\Install\Windows\Binary->download() C:\server\pickle\src\Package\PHP\Command\Install\Windows\Binary.php:328
PHP  10. file_get_contents() C:\server\pickle\src\Package\PHP\Command\Install\Windows\Binary.php:229

Warning: file_get_contents(): SSL operation failed with code 1. OpenSSL Error messages:
error:14095126:SSL routines:ssl3_read_n:unexpected eof while reading in C:\server\pickle\src\Package\PHP\Command\Install\Windows\Binary.php on line 229

Call Stack:
    0.0007     396008   1. {main}() C:\server\pickle\bin\pickle:0
    0.0259    1143096   2. Pickle\Console\Application->run() C:\server\pickle\bin\pickle:39
    0.1084    1425832   3. Pickle\Console\Application->doRun() C:\server\pickle\vendor\symfony\console\Application.php:140
    0.1237    1959360   4. Pickle\Console\Application->doRunCommand() C:\server\pickle\vendor\symfony\console\Application.php:264
    0.1237    1959360   5. Pickle\Console\Command\InstallerCommand->run() C:\server\pickle\vendor\symfony\console\Application.php:911
    0.1240    1961456   6. Pickle\Console\Command\InstallerCommand->execute() C:\server\pickle\vendor\symfony\console\Command\Command.php:258
    0.1248    1965440   7. Pickle\Console\Command\InstallerCommand->binaryInstallWindows() C:\server\pickle\src\Console\Command\InstallerCommand.php:200
    0.5675    2415464   8. Pickle\Package\PHP\Command\Install\Windows\Binary->install() C:\server\pickle\src\Console\Command\InstallerCommand.php:138
   10.4162    2416368   9. Pickle\Package\PHP\Command\Install\Windows\Binary->download() C:\server\pickle\src\Package\PHP\Command\Install\Windows\Binary.php:328
   10.4165    2417992  10. file_get_contents() C:\server\pickle\src\Package\PHP\Command\Install\Windows\Binary.php:229

 231863/231863 [============================] 100%Extracting archives...
copying C:\Users\Reza Salarmehr\AppData\Local\Temp\1\pickle-memcache\php_memcache.dll to C:\server\php\ext\php_memcache.dll
```

**Comment by pierrejoye** [comment]
Can you try using the phar pls?

Using master, you have to make sure that the cacert certifates bundles is available and loaded (as part of composer we use in vendor/composer/).

**Comment by cmb69** [comment]
This is an instance of https://bugs.php.net/79589. TL;DR: OpenSSL 1.1.1e introduced some very strict checks, which have been lifted as of OpenSSL 1.1.1f. So just update to PHP 7.4.6, and the problem should be solved.

**Comment by pierrejoye** [comment]
Yes, also ensure you have one these ini settings set:
openssl.cafile
openssl.capath
if needed. Composer's Ca-Bundle includes the latest available cacert.pem. 

Afair on Windows openssl still requires the cacert.pem, while curl can use the Windows Cert DB. @cmb69 do you remember the state of the ssl certs affair for openssl?

**Comment by pierrejoye** [comment]
@cmb69 btw this issue, together with Composer, does happen as well with 7.4.6. Expected if no cacert are loaded.

**Comment by salarmehr-cbs** [comment]
I used `https://curl.haxx.se/ca/cacert.pem` for `openssl.cafile`  and left`openssl.capath` blank, still getting the same error. Phar file returns a similar error 

@pierrejoye If you don't have a Win machine, maybe a docker image can be used to replicate the issue locally.

**Comment by salarmehr-cbs** [comment]
@pierrejoye Can you consider encouraging people to use phar file? the current readme file discourage it

**Comment by pierrejoye** [comment]
@salarmehr-cbs I modifed the README.md accordingly, promoting phar usage in the first place. Thanks for the headup :)

**Comment by pierrejoye** [comment]
@salarmehr-cbs do you get the error using phar or src? I get it using the src but not using phar. I have a windows machines, and a couple of VMs to test, thanks! 

And thanks again to contribute and test, much appreciated :)

**Comment by salarmehr-cbs** [comment]
@pierrejoye I get the same error using the phar file.

**Comment by pierrejoye** [comment]
`PS C:\Apps> .\php\php.exe .\pickle.phar info memcache
  - Installing memcache (latest-stable): Downloading (100%)
+-----------------------------------+----------+
| Package name                      | memcache |
| Package version (current release) | 4.0.5.2  |
| Package status                    | stable   |
+-----------------------------------+----------+

memcached extension

Configure options
+--------+--------------------+---------+
| Type   | Description        | Default |
+--------+--------------------+---------+
| enable | "memcache support" | yes     |
+--------+--------------------+---------+`

**Comment by pierrejoye** [comment]
`PS C:\Users\pierr\Documents\GitHub\pickle> C:\Apps\php\php.exe bin\pickle install memcache
+---------------+---------------------+
| php Path      | C:\Apps\php\php.exe |
| php Version   | 7.4.6               |
| Compiler      | vc15                |
| Architecture  | x64                 |
| Thread safety | no                  |
| Extension dir | C:\Apps\php\ext     |
| php.ini       | C:\Apps\php\php.ini |
+---------------+---------------------+
downloading https://windows.php.net/downloads/pecl/releases/memcache/4.0.5.2/php_memcache-4.0.5.2-7.4-nts-vc15-x64.zip
 231863/231863 [============================] 100%Extracting archives...
copying C:\Users\pierr\AppData\Local\Temp\pickle-memcache\php_memcache.dll to C:\Apps\php\ext\php_memcache.dll`

Fix in master for another possible issue about the progress bar. I am testing now more about SSL and let me upload a phar here (from master as well) in a few mins

**Comment by pierrejoye** [comment]
Using pickle.phar, built from master:

`PS C:\Apps> .\php\php.exe .\pickle.phar install memcache
+---------------+---------------------+
| php Path      | C:\Apps\php\php.exe |
| php Version   | 7.4.6               |
| Compiler      | vc15                |
| Architecture  | x64                 |
| Thread safety | no                  |
| Extension dir | C:\Apps\php\ext     |
| php.ini       | C:\Apps\php\php.ini |
+---------------+---------------------+
downloading https://windows.php.net/downloads/pecl/releases/memcache/4.0.5.2/php_memcache-4.0.5.2-7.4-nts-vc15-x64.zip
 231863/231863 [============================] 100%Extracting archives...
copying C:\Users\pierr\AppData\Local\Temp\pickle-memcache\php_memcache.dll to C:\Apps\php\ext\php_memcache.dll`

Using pickle.phar built from master (attached here).
[pickle.zip](https://github.com/FriendsOfPHP/pickle/files/4729001/pickle.zip)

May I ask you try this phar on your windows?

**Comment by cmb69** [comment]
> Afair on Windows openssl still requires the cacert.pem, while curl can use the Windows Cert DB.

I think curl would have to be build with Schannel support to use the Windows Cert DB, but official builds are with OpenSSL support.

**Comment by pierrejoye** [comment]
@cmb69 thanks for you hint the other day, that fixes the notification call back :)

**Comment by salarmehr-cbs** [comment]
@pierrejoye I tried yr new file doesn't work.

**Comment by pierrejoye** [comment]
@salarmehr-cbs weird. I am not sure how to move forwards. Do you get the same error as above?

I tried on three different windows (clean fresh ones too) and it works as expected.

Are you still using 7.4.5? Maybe could you try with 7.4.6? 

And maybe show me your php -m and php -i output pls.

**Comment by cmb69** [comment]
For me, the pickle.phar from [above](https://github.com/FriendsOfPHP/pickle/issues/175#issuecomment-638738154) works fine with a fresh PHP 7.4.6, but with PHP 7.4.5 I get the error ("SSL routines:ssl3_read_n:unexpected eof while reading"). Same OpenSSL setup, but different versions (1.1.1e vs. 1.1.1g).

**Comment by pierrejoye** [comment]
@cmb69 thanks! Looks like the strictness being fixed in 1.1.1g you mentioned last week.

@salarmehr-cbs if you are OK, I would like to close this ticket, not much we can do on the pickle/composer side.

**Comment by pierrejoye** [comment]
I will close this issue for now. Please re open if you have any issue. :)
